### DISCUSSION TITLE
Redis optimization: avoid querying Redis when threshold is violated

### DISCUSSION THREAD

**Discussion by oleg-smith** [vladimir-bukhtoyarov/bucket4j#384] [General]
Hi, I would like to verify Redis integration behavior for this scenario:

- a bucket backed by Redis
- the bucket bandwidth is 50 requests/min, classic
- say, all 50 tokens are consumed in the very first second of a minute.

1. Would the library query Redis for the rest of the minute, knowing the threshold had been already violated?

2. If it will query Redis by default, is there a way to configure it not to?

Thanks!

**Comment by vladimir-bukhtoyarov** [comment]
Hello @oleg-smith 

Refill is not a single source of increasing tokens in the bucket. There are API that can to increasing tokens in the bucket:
* [addTokens](https://github.com/bucket4j/bucket4j/blob/8.3/bucket4j-core/src/main/java/io/github/bucket4j/Bucket.java#L155)
* [forceAddTokens](https://github.com/bucket4j/bucket4j/blob/8.3/bucket4j-core/src/main/java/io/github/bucket4j/Bucket.java#L177)
* [reset](https://github.com/bucket4j/bucket4j/blob/8.3/bucket4j-core/src/main/java/io/github/bucket4j/Bucket.java#L182)
* [replaceConfiguration](https://github.com/bucket4j/bucket4j/blob/8.3/bucket4j-core/src/main/java/io/github/bucket4j/Bucket.java#L259)

So, proposed optimization can not be applied by default because library does not know the scope of API that user intents to call. Anyway, this kind of optimization can be configured expilitly. There are big package of library that is not documented yet https://github.com/bucket4j/bucket4j/tree/8.4/bucket4j-core/src/main/java/io/github/bucket4j/distributed/proxy/optimization this package is addressed to implement varios optimization strategeis, feel free to read javadocs for classes in this packet, it can be interesting. 

I have just added optimization for your use case in this commit https://github.com/bucket4j/bucket4j/commit/d1e4d7520807631bdcc8fbf1abc51a272eba0563   The class HazelcastSkipSyncOnZeroExample shows how to use this optimization:
![image](https://github.com/bucket4j/bucket4j/assets/6761830/f88e6f74-fea0-46ee-8035-005652aa89cd)
The ```8.4.0``` is going to be released at Jule, so just copy and past my code to your project if you have no time to wait.

**Reply by oleg-smith** [reply]
Thanks very much!

**Comment by oleg-smith** [comment]
Hi @vladimir-bukhtoyarov, I tried it this way:

```
remoteProxyManager = LettuceBasedProxyManager
           ...
            .build();
```
            
Then, accessing the same bucket from multiple threads this way:

```
var optimizationListener = new DefaultOptimizationListener();
        var optimization = new SkipSyncOnZeroOptimization(optimizationListener, TimeMeter.SYSTEM_NANOTIME)
            .withListener(optimizationListener);

        return remoteProxyManager.builder()
            .withOptimization(optimization)
            .build(
                key.getBytes(),
                () -> BucketConfiguration.builder()
                    .addLimit(Bandwidth.classic(capacity, Refill.intervally(capacity, period)))
                    .build()
            );
```
            
            
in debug, I'm seeing `isNeedToExecuteRemoteImmediately` always returns true:

<img width="1070" alt="image" src="https://github.com/bucket4j/bucket4j/assets/6909571/ded0a2b1-de91-4b70-87c6-ac5217732949">


Also, in the debug, the optimization object is always a _new_ object the way I use it above - I assume that is the issue.



- Should I cache the optimization object additionally by the bucket key using CHM, for instance?

- Is optimization object thread-safe by contract?

- Could you please suggest a better way for configuring an optimization and preserving its state, when accessed from multiple threads?

**Reply by vladimir-bukhtoyarov** [reply]
Hello @oleg-smith 

> Should I cache the optimization object additionally by the bucket key using CHM, for instance?

You should to cache the whole bucket, see this javadocs https://github.com/bucket4j/bucket4j/blob/8.3/bucket4j-core/src/main/java/io/github/bucket4j/distributed/proxy/RemoteBucketBuilder.java#L54

> Is optimization object thread-safe by contract?

Yes, it is thread-safe

> Could you please suggest a better way for configuring an optimization and preserving its state, when accessed from multiple threads?

As mentioneed above, in oposite to nob-optimized buckets, the the optimized buckets should be cached in some way, choose your own.

**Reply by oleg-smith** [reply]
ok, I switched to caching whole bucket with CHM

This leads to a case where bucket is filled once and never refilled:


`buckets chm<String, Bucket>;`

use:

`buckets.computeIfAbsent(key, … same init using lettuceProxyManager..).tryConsume(1)`


Result: 5 tokens/minute:

allowed
allowed
allowed
allowed
allowed
not allowed
not allowed
… never allowed after that

**Reply by vladimir-bukhtoyarov** [reply]
It is not clear how this log was produced.

**Reply by oleg-smith** [reply]
`System.out.println(tryConsume(1) ? "allowed" : "not allowed")`

**Reply by vladimir-bukhtoyarov** [reply]
The whole code

**Reply by vladimir-bukhtoyarov** [reply]
I have fixed the example. TimeMeter.SYSTEM_MILLISECONDS should be used

![image](https://github.com/bucket4j/bucket4j/assets/6761830/5c98cc86-d94a-46b0-bde7-8ce53680d7f1)

**Reply by oleg-smith** [reply]
thanks, let me try!

Btw, isn't it enough to use `new SkipSyncOnZeroOptimization(optimizationListener, TimeMeter.SYSTEM_MILLISECONDS)`?

without `.withListener()` part

**Reply by vladimir-bukhtoyarov** [reply]
Yes it is enough. Listener is needed only to collect statistics, nothing more.

**Reply by oleg-smith** [reply]
Yeah, I meant that the listener is passed 2 times - 1st in the constructor, and 2nd in the `withListener`, basically making a copy 

that just seemed excessive

**Comment by vladimir-bukhtoyarov** [comment]
released with version 8.4.0
